File: fathersDay/Dsa/Day4 (237. Delete Node in a Linked List)/Solution.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def DeleteNode(self, node: ListNode) -> None:
        if node is None or node.next is None:
            return 
        node.val = node.next.val
        node.next = node.next.next
        
solution = Solution()
head = ListNode(4, ListNode(5, ListNode(1, ListNode(9))))
node_to_delete = head.next  
solution.DeleteNode(node_to_delete)
current = head
while current:
    print(current.val, end=" -> ")
    current = current.next
print("None")  

File: fathersDay/Dsa/Day4 (237. Delete Node in a Linked List)/Solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(nullptr) {}
 * };
 */
class Solution {
public:
    void deleteNode(ListNode* node) {
        // According to the problem constraints, the given node will not be the tail node
        // and will always be a valid node in the list.
        // However, for robustness, we can add checks:
        if (node == nullptr || node->next == nullptr) {
            // Cannot delete if node is null or it's the last node
            // (as we need to copy from node->next)
            return;
        }
        
        // Copy the value of the next node to the current node
        node->val = node->next->val;
        
        // Store the node to be deleted (the next node)
        ListNode* nodeToDelete = node->next;
        
        // Link the current node to the node after the one being deleted
        node->next = node->next->next;
        
        // Free the memory of the deleted node (important in C++)
        delete nodeToDelete;
    }
};